### ISSUE TITLE
[Bug] KDA context-parallel backward: gk loads miss head/bos offset in bwd pre-process (wrong grads on non-last ranks)

### ISSUE THREAD

**Issue by dhnihaoya** [fla-org/flash-linear-attention#1115]
> **Update (2026-08-09, later the same day): already fixed on `main` and in release v0.5.2.** The kernel rewrite in 02e270330b ("[Ops] Migrate kernels off tl.make_block_ptr / tl.advance for triton main compat", 2026-07-25) added the missing pointer advance (`gk += ((bos * HV + i_h) * K)`), and v0.5.2 (2026-07-27) contains it. So the bug affects **0.4.2 and any build older than 2026-07-25** — which is what we had pinned. My earlier claim that `main` was equally affected came from a stale mirror snapshot (5da31d1, actually 2026-03-11) — apologies for the noise. The analysis below still applies to the affected versions; feel free to close, or keep for a regression test.

## Summary

When running `chunk_kda` with `cp_context` (context parallelism), the **backward pass produces wrong input
gradients on every non-last CP rank** (~10% relative error on dk/dv/dg/dbeta in fp32), while the forward pass
is exact and the last rank's gradients are exact.

Root cause: in `fla/ops/cp/chunk_delta_h.py`, the dh-recurrence section of `pre_process_bwd_kernel_merged`
loads the per-channel gate `gk` without the head/batch-sequence offset — the pointer advance
`gk += (bos * H + i_h) * K` that the single-card reference kernel
(`fla/ops/common/chunk_delta_h.py`, `chunk_gated_delta_rule_bwd_kernel_dhu_blockdim64`, lines 405-406)
applies was lost when the loop body was copied over. As a result, `b_dh *= exp2(b_gk_last)` uses **head 0's
gates for every head**, so the cross-rank state gradient (`dhm` → `dht`) is wrong for all heads except head 0,
and every rank before the last inherits the error.

## Environment

- fla 0.4.2 (pip). **Affected: 0.4.2 and builds older than 2026-07-25. Fixed: `main` since 02e270330b (2026-07-25) and release v0.5.2** (see top note).
- torch 2.12.1, 2× A100 80GB, CP degree 2

## Symptom (minimal repro, fp32, exact arithmetic)

`torchrun --nproc_per_node=2 check_fla_kcp.py` — compares `chunk_kda` one-shot vs CP×2 contiguous shards
(L=8192, H=4, K=V=128, fp32, `use_qk_l2norm_in_kernel=True`, `use_gate_in_kernel=True`,
`transpose_state_layout=True`); one-shot grads from the same op without `cp_context`:

| quantity | rank 0 (needs cross-rank state bwd) | rank 1 (last rank) |
|---|---|---|
| forward o | 1.6e-5 (exact) | — |
| dq | 0.0 | 1.8e-4 |
| dk | **1.36e-1 rel** | 2.4e-4 |
| dv | **1.15e-1 rel** | 0.0 |
| dg | **1.04e-1 rel** | 1.5e-4 |
| dbeta | **9.5e-2 rel** | 5.3e-5 |
| dA_log / ddt_bias | 1.0e-2 / 2.1e-2 | — |

The error profile localizes exactly to the cross-rank state-gradient path: forward exact, last rank exact,
dq exact (dq does not depend on `dht`), and only the quantities that backprop through the merged state
gradient are wrong. We also reproduced this under the official test setup (bf16, `safe_gate=True`,
`naive_recurrent_kda` reference): CP mode exceeds the official tolerance (`ratio=8e-3`) by 10-80× on
dk/dv/dg/db, while single-card `chunk_kda` on the same inputs sits at 1-4%.

## Root cause

In `pre_process_bwd_kernel_merged` (lines 956-991), the dh section loads:

```python
b_gk_last1 = tl.load(gk + last_idx * H * K + o_k1, ...)   # missing (bos ... ) and i_h * K !
```

The reference kernel advances the pointer once before the loop (`gk += (bos * H + i_h) * K`), making the same
addressing correct; the CP copy dropped that line. The forward kernels in the same file
(e.g. lines 400-423, 501) and even the dm section of the same backward kernel (line 1111) all use the correct
`(bos + last_idx) * H * K + i_h * K + ...` form — only the backward dh section is wrong.

The same defect exists in the (currently unused) `pre_process_bwd_kernel_stage1` (lines 761-785).

Impact: CP + KDA (`USE_GK=True`) with H > 1 → wrong grads on all non-last ranks. H = 1 with a single sequence
is accidentally unaffected (the wrong address equals the right one); varlen with `bos ≠ 0` is affected even at
H = 1.

## Fix

12 load sites, see attached `fla_kcp_fix.patch`:

```diff
- b_gk_lastN = tl.load(gk + last_idx * H * K + o_kN, ...)
+ b_gk_lastN = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_kN, ...)
```

With the patch applied, the repro above **passes**: rank-0 rel errors on dk/dv/dbeta drop to **0.0** and dg to **9.8e-8** (from 1.36e-1 / 1.15e-1 / 9.5e-2 / 1.04e-1); dA_log/ddt_bias drop to 4.5e-5 / 7.2e-5 (from 1.0e-2 / 2.1e-2); forward unchanged at 1.6e-5 — i.e. CP backward now matches one-shot `chunk_kda` to fp32 tolerance (verified 2026-08-09, 2× A100).

## Side observations (not part of this fix)

- The `USE_G` branch of the same kernel ignores `USE_EXP2` and always uses `exp` (lines 937-944), while the
  reference kernel uses `exp2` when `USE_EXP2` — possibly another latent issue for GDN + CP + exp2.
- Forward passes `cu_seqlens[-2:]` while backward passes `cu_seqlens[:2]` (lines 1176 / 1261); identical for a
  single sequence but asymmetric for multi-sequence varlen batches.

## Repro script & patch

<details><summary>check_fla_kcp.py (minimal repro, pure fla API)</summary>

```python
"""Pure-fla KCP check: chunk_kda one-shot vs CP×2 (cp_context), fwd + input grads.

No phageseek code involved — isolates whether fla's KCP (fla.ops.cp) itself is
exact in this install. Random q/k/v/g/beta, L=8192, H=4, K=V=128, fp32.

Run: torchrun --nproc_per_node=2 check_fla_kcp.py
"""
import sys

import torch
import torch.distributed as dist
from torch.distributed.device_mesh import init_device_mesh

from fla.ops.cp import build_cp_context
from fla.ops.kda import chunk_kda

L, H, K = 8192, 4, 128

dist.init_process_group("nccl")
rank, world = dist.get_rank(), dist.get_world_size()
torch.cuda.set_device(rank)
dev = torch.device("cuda", rank)
torch.manual_seed(42)

mesh = init_device_mesh("cuda", (world,), mesh_dim_names=("cp",))
group = mesh["cp"].get_group()
c = L // world


def make_inputs():
    q = torch.randn(1, L, H, K, device=dev) * 0.5
    k = torch.randn(1, L, H, K, device=dev) * 0.5
    v = torch.randn(1, L, H, K, device=dev) * 0.5
    g_raw = torch.randn(1, L, H, K, device=dev) * 0.1
    beta = torch.rand(1, L, H, device=dev) * 0.9 + 0.05
    A_log = torch.log(torch.empty(H, device=dev).uniform_(1.0, 16.0))
    dt = torch.exp(torch.rand(H * K, device=dev) * 4.6 - 7.0).clamp(min=1e-4)
    dt_bias = dt + torch.log(-torch.expm1(-dt))
    return q, k, v, g_raw, beta, A_log, dt_bias


# ── one-shot reference (single card, full sequence) ─────────────────────────
q, k, v, g_raw, beta, A_log, dt_bias = make_inputs()
for t in (q, k, v, g_raw, beta):
    t.requires_grad_(True)
A_log.requires_grad_(True)
dt_bias.requires_grad_(True)
ref = (q, k, v, g_raw, beta, A_log, dt_bias)
o_ref, _ = chunk_kda(
    q, k, v, g_raw, beta,
    A_log=A_log, dt_bias=dt_bias,
    use_qk_l2norm_in_kernel=True, use_gate_in_kernel=True,
    output_final_state=False, transpose_state_layout=True,
)
o_ref.sum().backward()
g_ref = {n: t.grad.detach().clone() for n, t in zip("q k v g beta A_log dt_bias".split(), ref)}

# ── CP×2: contiguous shard + cp_context ─────────────────────────────────────
q2, k2, v2, g2, b2, A2, d2 = (t.detach().clone() for t in ref)
for t in (q2, k2, v2, g2, b2):
    t.requires_grad_(True)
A2.requires_grad_(True)
d2.requires_grad_(True)
cp_tensors = (q2, k2, v2, g2, b2, A2, d2)

cu_global = torch.tensor([0, L], dtype=torch.long, device=dev)
ctx = build_cp_context(cu_global, group=group, conv1d_kernel_size=4)
sl = slice(rank * c, (rank + 1) * c)
o_cp, _ = chunk_kda(
    q2[:, sl], k2[:, sl], v2[:, sl], g2[:, sl], b2[:, sl],
    A_log=A2, dt_bias=d2,
    use_qk_l2norm_in_kernel=True, use_gate_in_kernel=True,
    output_final_state=False, transpose_state_layout=True,
    cp_context=ctx,
)
o_cp.sum().backward()

# gather outputs + input grads
o_all = [torch.empty_like(o_cp) for _ in range(world)]
dist.all_gather(o_all, o_cp.detach())
o_full = torch.cat(o_all, dim=1)

ok = True
if rank == 0:
    d = (o_full - o_ref).abs().max().item()
    print(f"[fwd] max|diff| = {d:.3e}")
    if d > 1e-4:
        print("FAIL fwd")
        ok = False
for n, t, tref in zip("q k v g beta A_log dt_bias".split(), cp_tensors, ref):
    g = t.grad.detach().clone()
    if g.dim() >= 3:  # sequence tensor: compare per-rank halves BEFORE summing
        rg = g_ref[n]
        for r in range(world):
            sl = slice(r * c, (r + 1) * c)
            gr = torch.zeros_like(g)
            dist.broadcast(gr, src=r) if False else None
        # gather each rank's local-grad tensor and compare its OWN half only
        allg = [torch.empty_like(g) for _ in range(world)]
        dist.all_gather(allg, g)
        if rank == 0:
            for r in range(world):
                sl = slice(r * c, (r + 1) * c)
                d0 = (allg[r][:, sl] - rg[:, sl]).abs().max().item()
                s = rg.abs().max().item()
                print(f"[grad] {n:6s} rank{r}-half max|diff|={d0:.3e} (max|grad|={s:.3e}) rel={d0/max(s,1e-12):.2e}")
        continue
    dist.all_reduce(g, op=dist.ReduceOp.SUM, group=group)
    rg = g_ref[n]
    d = (g - rg).abs().max().item()
    s = rg.abs().max().item()
    if rank == 0:
        print(f"[grad] {n:8s} max|diff|={d:.3e} (max|grad|={s:.3e}) rel={d/max(s,1e-12):.2e}")
    if d > 1e-3 * max(1.0, s):
        ok = False
if rank == 0:
    print("PASS" if ok else "FAIL (informational run)")
dist.destroy_process_group()
sys.exit(0 if ok else 1)

```

</details>

<details><summary>fla_kcp_fix.patch (12 load sites)</summary>

```diff
--- a/fla/ops/cp/chunk_delta_h.py
+++ b/fla/ops/cp/chunk_delta_h.py
@@ -758,7 +758,7 @@
         b_k = tl.load(p_k, boundary_check=(0, 1))
         if USE_GK:
             o_k1 = tl.arange(0, 64)
-            b_gk_last1 = tl.load(gk + last_idx * H*K + o_k1, mask=(o_k1 < K), other=0.).to(tl.float32)
+            b_gk_last1 = tl.load(gk + (bos + last_idx) * H*K + i_h * K + o_k1, mask=(o_k1 < K), other=0.).to(tl.float32)
         b_dv = tl.dot(b_k, b_dh1.to(b_k.dtype))
 
         if K > 64:
@@ -766,7 +766,7 @@
             b_k = tl.load(p_k, boundary_check=(0, 1))
             if USE_GK:
                 o_k2 = 64 + o_k1
-                b_gk_last2 = tl.load(gk + last_idx * H*K + o_k2, mask=(o_k2 < K), other=0.).to(tl.float32)
+                b_gk_last2 = tl.load(gk + (bos + last_idx) * H*K + i_h * K + o_k2, mask=(o_k2 < K), other=0.).to(tl.float32)
             b_dv += tl.dot(b_k, b_dh2.to(b_k.dtype))
 
         if K > 128:
@@ -774,7 +774,7 @@
             b_k = tl.load(p_k, boundary_check=(0, 1))
             if USE_GK:
                 o_k3 = 128 + o_k1
-                b_gk_last3 = tl.load(gk + last_idx * H*K + o_k3, mask=(o_k3 < K), other=0.).to(tl.float32)
+                b_gk_last3 = tl.load(gk + (bos + last_idx) * H*K + i_h * K + o_k3, mask=(o_k3 < K), other=0.).to(tl.float32)
             b_dv += tl.dot(b_k, b_dh3.to(b_k.dtype))
 
         if K > 192:
@@ -782,7 +782,7 @@
             b_k = tl.load(p_k, boundary_check=(0, 1))
             if USE_GK:
                 o_k4 = 192 + o_k1
-                b_gk_last4 = tl.load(gk + last_idx * H*K + o_k4, mask=(o_k4 < K), other=0.).to(tl.float32)
+                b_gk_last4 = tl.load(gk + (bos + last_idx) * H*K + i_h * K + o_k4, mask=(o_k4 < K), other=0.).to(tl.float32)
             b_dv += tl.dot(b_k, b_dh4.to(b_k.dtype))
 
         if USE_G:
@@ -953,9 +953,9 @@
             if USE_GK:
                 o_k1 = tl.arange(0, 64)
                 if USE_EXP2:
-                    b_gk_last1 = tl.load(gk + last_idx * H * K + o_k1, mask=(o_k1 < K), other=0.).to(tl.float32)
+                    b_gk_last1 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k1, mask=(o_k1 < K), other=0.).to(tl.float32)
                 else:
-                    b_gk_last1 = tl.load(gk + last_idx * H * K + o_k1, mask=(o_k1 < K), other=0.).to(tl.float32)
+                    b_gk_last1 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k1, mask=(o_k1 < K), other=0.).to(tl.float32)
             b_dv = tl.dot(b_k, b_dh1.to(b_k.dtype))
 
             if K > 64:
@@ -964,9 +964,9 @@
                 if USE_GK:
                     o_k2 = 64 + o_k1
                     if USE_EXP2:
-                        b_gk_last2 = tl.load(gk + last_idx * H * K + o_k2, mask=(o_k2 < K), other=0.).to(tl.float32)
+                        b_gk_last2 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k2, mask=(o_k2 < K), other=0.).to(tl.float32)
                     else:
-                        b_gk_last2 = tl.load(gk + last_idx * H * K + o_k2, mask=(o_k2 < K), other=0.).to(tl.float32)
+                        b_gk_last2 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k2, mask=(o_k2 < K), other=0.).to(tl.float32)
                 b_dv += tl.dot(b_k, b_dh2.to(b_k.dtype))
 
             if K > 128:
@@ -975,9 +975,9 @@
                 if USE_GK:
                     o_k3 = 128 + o_k1
                     if USE_EXP2:
-                        b_gk_last3 = tl.load(gk + last_idx * H * K + o_k3, mask=(o_k3 < K), other=0.).to(tl.float32)
+                        b_gk_last3 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k3, mask=(o_k3 < K), other=0.).to(tl.float32)
                     else:
-                        b_gk_last3 = tl.load(gk + last_idx * H * K + o_k3, mask=(o_k3 < K), other=0.).to(tl.float32)
+                        b_gk_last3 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k3, mask=(o_k3 < K), other=0.).to(tl.float32)
                 b_dv += tl.dot(b_k, b_dh3.to(b_k.dtype))
 
             if K > 192:
@@ -986,9 +986,9 @@
                 if USE_GK:
                     o_k4 = 192 + o_k1
                     if USE_EXP2:
-                        b_gk_last4 = tl.load(gk + last_idx * H * K + o_k4, mask=(o_k4 < K), other=0.).to(tl.float32)
+                        b_gk_last4 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k4, mask=(o_k4 < K), other=0.).to(tl.float32)
                     else:
-                        b_gk_last4 = tl.load(gk + last_idx * H * K + o_k4, mask=(o_k4 < K), other=0.).to(tl.float32)
+                        b_gk_last4 = tl.load(gk + (bos + last_idx) * H * K + i_h * K + o_k4, mask=(o_k4 < K), other=0.).to(tl.float32)
                 b_dv += tl.dot(b_k, b_dh4.to(b_k.dtype))
 
             if USE_G:

```

</details>

**Comment by dhnihaoya** [comment]
Correction / good news: while preparing the PR I found this is **already fixed on `main`** — the 2026-07-25 kernel rewrite 02e270330b (triton-main compatibility migration) rewrote these kernels with the correct `gk += ((bos * HV + i_h) * K)` advance, and **v0.5.2** (2026-07-27) already ships it. Verified by code inspection of `fla/ops/cp/chunk_delta_h.py` at 9c8e42e7 and 7843b328b0 (advance present at both, buggy pattern gone). My initial "main is also affected" claim was based on a stale mirror clone (5da31d1 = 2026-03-11) — sorry about that.

Net: anyone on **0.4.2 or builds older than 2026-07-25** using KDA + `cp_context` with H>1 gets silently wrong backward grads (~10% on dk/dv/dg/dβ, non-last ranks). Upgrading to ≥ v0.5.2 fixes it. The fp32 repro + root-cause analysis above may still be useful as a regression test (CP×2 vs one-shot parity). No PR from our side since `main` no longer needs the patch.
